File: quatPropagate.m
function possNewQuats = quatPropagate(possQuats,possAngV,gyroDt)
    %
    % quatPropogate(possQuats,possAngV,gyroDt)
    % finds the possible orientations of the spacecraft after the possible
    %   rotations of the spacecraft
    %
    % Parameters:
    % possQuats is the array of pre-propogation sigma
    %   quaternions
    % possAngV is the array of estimated sigma angular
    %   velocities (measured minus Chi bias sigma vectors)
    % gyroDt is the sampling interval in the gyro
    %
    % Results:
    % possNewQuats is the array of propagated sigma quaternions
    
    possNewQuats = zeros(4,size(possAngV,2));
    for i=1:size(possAngV,2)
        
        % Eq. (29)
        psiK = ((sin(0.5*norm(possAngV(:,i))*gyroDt))*possAngV(:,i))/...
                norm(possAngV(:,i));
        Omega = [cos(0.5*norm(possAngV(:,i))*gyroDt)*eye(3)-crossMatrix(...
            psiK), psiK; -psiK', cos(0.5*norm(possAngV(:,i))*gyroDt)];
        
        possNewQuats(:,i) = Omega*possQuats(:,i);
    end
end

File: src/usque/functions/quatPropagate.cpp
#include "usque.hpp"

namespace RPL {
namespace USQUE {

Matrix_4x13d quatPropagate(
	Matrix_4x13d& possQuats, 
	Matrix_3x13d& possAngV, 
	double gyroDt
) {
	Matrix_4x13d possNewQuats;
	for(int i = 0; i < 2 * SIZE + 1; i++) {
		Eigen::Vector3d psiK = sin(0.5*gyroDt*possAngV.col(i).norm())*
			possAngV.col(i)/possAngV.col(i).norm();
		Eigen::Matrix4d omega;
		omega.block(0,0,3,3) << cos(0.5*gyroDt*possAngV.col(i).norm())*
			Eigen::Matrix3d::Identity() - crossMatrix(psiK);
		omega.block(0,3,3,1) << psiK;
		omega.row(3) << -1*psiK.transpose(), cos(0.5*gyroDt*possAngV.col(i).
			norm());
		possNewQuats.col(i) << omega*possQuats.col(i);
	}
	return possNewQuats;
}

}
}